src/dbnode/storage/series/buffer.go

```diff
@@ -820,59 +987,118 @@ func (b *dbBufferBucket) merge() (mergeResult, error) {
+	return merges, nil
 }
 
-type discardMergedResult struct {
-	block  block.DatabaseBlock
-	merges int
+// mergeStreamsToEncoder merges streams to an encoder and returns the last
+// write time. It is the responsibility of the caller to close the returned
+// encoder when appropriate.
+func mergeStreamsToEncoder(
+	blockStart time.Time,
+	streams []xio.SegmentReader,
+	opts Options,
+) (encoding.Encoder, time.Time, error) {
+	bopts := opts.DatabaseBlockOptions()
+	encoder := opts.EncoderPool().Get()
+	encoder.Reset(blockStart, bopts.DatabaseBlockAllocSize())
+	iter := bopts.MultiReaderIteratorPool().Get()
+	defer func() {
+		iter.Close()
+	}()
+
+	var lastWriteAt time.Time
+	iter.Reset(streams, blockStart, opts.RetentionOptions().BlockSize())
+	for iter.Next() {
+		dp, unit, annotation := iter.Current()
+		if err := encoder.Encode(dp, unit, annotation); err != nil {
+			encoder.Close()
+			return nil, timeZero, err
+		}
+		lastWriteAt = dp.Timestamp
+	}
+	if err := iter.Err(); err != nil {
+		encoder.Close()
+		return nil, timeZero, err
+	}
+
+	return encoder, lastWriteAt, nil
 }
 
-func (b *dbBufferBucket) discardMerged() (discardMergedResult, error) {
+func (b *BufferBucket) toStream(ctx context.Context) (xio.SegmentReader, error) {
 	if b.hasJustSingleEncoder() {
-		// Already merged as a single encoder
-		encoder := b.encoders[0].encoder
-		newBlock := b.opts.DatabaseBlockOptions().DatabaseBlockPool().Get()
-		blockSize := b.opts.RetentionOptions().BlockSize()
-		newBlock.Reset(b.start, blockSize, encoder.Discard())
-
-		// The single encoder is already discarded, no need to call resetEncoders
-		// just remove it from the list of encoders
-		b.encoders = b.encoders[:0]
 		b.resetBootstrapped()
-
-		return discardMergedResult{newBlock, 0}, nil
+		// Already merged as a single encoder.
+		return b.encoders[0].encoder.Stream(), nil
 	}
 
 	if b.hasJustSingleBootstrappedBlock() {
-		// Already merged just a single bootstrapped block
-		existingBlock := b.bootstrapped[0]
-
 		// Need to reset encoders but do not want to finalize the block as we
-		// are passing ownership of it to the caller
+		// are passing ownership of it to the caller.
 		b.resetEncoders()
-		b.bootstrapped = nil
-
-		return discardMergedResult{existingBlock, 0}, nil
+		return b.bootstrapped[0].Stream(ctx)
 	}
 
-	result, err := b.merge()
+	_, err := b.merge()
 	if err != nil {
 		b.resetEncoders()
 		b.resetBootstrapped()
-		return discardMergedResult{}, err
+		return nil, err
 	}
 
-	merged := b.encoders[0].encoder
+	// After a successful merge, encoders and bootstrapped blocks will be
+	// reset, and the merged encoder appended as the only encoder in the
+	// bucket.
+	if !b.hasJustSingleEncoder() {
+		return nil, errIncompleteMerge
+	}
 
-	newBlock := b.opts.DatabaseBlockOptions().DatabaseBlockPool().Get()
-	blockSize := b.opts.RetentionOptions().BlockSize()
-	newBlock.Reset(b.start, blockSize, merged.Discard())
+	return b.encoders[0].encoder.Stream(), nil
```

Comment:
Same here - need to register finalizer here from the context so that the copied segment will get returned.

e.g.
```go
stream := b.encoders[0].encoder.Stream()
ctx.RegisterFinalizer(stream)
return stream, nil
```

Notice that for the bootstrap case because it takes a context on the call to stream it will do the register finalizer step for you, but here we have to do it ourselves since we're dealing with just an encoder.

Reply:
Oh I see, thanks for explaining that.